python/foglamp/services/core/server.py

```diff
@@ -356,6 +361,87 @@ def _start_app(cls, loop, app, host, port, ssl_ctx=None):
         server = loop.run_until_complete(coro)
         return server, handler
 
+    """ Get the full path of FogLAMP PID file """
+    @staticmethod
+    def pid_filename():
+        if _FOGLAMP_DATA is None:
+            path = _FOGLAMP_ROOT + "/data"
+        else:
+            path = _FOGLAMP_DATA
+        return path + _FOGLAMP_PID_DIR + "/" + _FOGLAMP_PID_FILE 
+
+    """ Check whether the PID file exists """
+    @classmethod
+    def _pidfile_exists(cls):
+        try:
+            fh = open(cls._pidfile,'r')
+            fh.close()
+            return True
+        except (FileNotFoundError, IOError, TypeError):
+            return False
+
+    """ Remove PID file """
+    @classmethod
+    def _remove_pid(cls):
+        try:
+            os.remove(cls._pidfile)
+            _logger.info("FogLAMP PID file [" + cls._pidfile + "] removed.")
+        except Exception as ex:
+            _logger.error("FogLAMP PID file remove error: [" + ex.__class__.__name__ + "], (" + format(str(ex)) + ")")
+
+    """ Write data into PID file """
+    @classmethod
+    def _write_pid(cls, api_address, api_port):
+        try:
+            """ Get PID file path """
+            cls._pidfile = cls.pid_filename()
+
+            """ Check for existing PID file and log a message """
+            if cls._pidfile_exists() is True:
+                _logger.warn("A FogLAMP PID file has been found: [" + \
+                             cls._pidfile + "] found, ignoring it.")
+
+            """ Get the running script PID """
+            pid = os.getpid()
+
+            """ Open for writing and truncate existing file """
+            fh = None
+            try:
+                fh = open(cls._pidfile, 'w+')
+            except FileNotFoundError:
+                try:
+                    os.makedirs(os.path.dirname(cls._pidfile))
+                    _logger.info("The PID directory [" + os.path.dirname(cls._pidfile) + "] has been created")
+                    fh = open(cls._pidfile, 'w+')
+                except Exception as ex:
+                    errmsg = "PID dir create error: [" + ex.__class__.__name__ + "], (" + format(str(ex)) + ")"
+                    _logger.error(errmsg)
+                    sys.stderr.write(errmsg + "\n")
+                    raise
+            except Exception as ex:
+                errmsg = "FogLAMP PID file create error: [" + ex.__class__.__name__ + "], (" + format(str(ex)) + ")"
+                _logger.error(errmsg)
+                sys.stderr.write(errmsg + "\n")
+                raise
+
+            """ Build the JSON object to write into PID file """
+            info_data = {'processID' : pid,\
+                         'REST_info' : {\
+                             "protocol": "HTTP" if cls.is_rest_server_http_enabled else "HTTPS",\
+                             "address": api_address,\
```

Comment:
I would be tempted to make the address a list for future expansion. That way when we want to support binding to multiple interfaces we can. The current 0.0.0.0 approach is not great as this is not really the address we listen on, but the address we bind to and the meaning is subtly different.